File: dxDrawLib/src/Client/dxdrawlib/helpers/Color.ts
/// <reference path="../../dxdrawlib.d.ts" />

class Color {

	private _a: number;
	private _r: number;
	private _g: number;
	private _b: number;

	set a(value: number) { this._a = Math.round(clamp(value, 0, 255)); }
	get a(): number { return this._a; }
	set r(value: number) { this._r = Math.round(clamp(value, 0, 255)); }
	get r(): number { return this._r; }
	set g(value: number) { this._g = Math.round(clamp(value, 0, 255)); }
	get g(): number { return this._g; }
	set b(value: number) { this._b = Math.round(clamp(value, 0, 255)); }
	get b(): number { return this._b; }

	constructor(a: number|string = 200, r: number = 0, g: number = 0, b: number = 0) {
		if(typeof a == "number") {
			this.a = a;
			this.r = r;
			this.g = g;
			this.b = b;
		} else if(typeof a == "string") {	
			if(a.charAt(0) == '#') a = a.substr(1);
			let value = parseInt(a, 16);
			
			if (a.length == 8) this._a = (value >> 8 * 3) & 255;
			else this._a = 255;

			this._r = (value >> 8 * 2) & 255;
			this._g = (value >> 8) & 255;
			this._b = value & 255;
		}		
	}
	
	public set(a: number, r: number, g: number, b: number): void {
		this.a = a;
		this.r = r;
		this.g = g;
		this.b = b;
	}

	public Lighten(factor: number): Color
	{
		this.r = (this._r + (255 - this._r) * factor);
		this.g = (this._g + (255 - this._g) * factor);
		this.b = (this._b + (255 - this._b) * factor);
		return this;
	}
	
	public Darken(factor: number): Color
	{
		this.r = this._r * (1 - factor);
		this.g = this._g * (1 - factor);
		this.b = this._b * (1 - factor);
		return this;
	}
	
	public Clone(): Color {
		return new Color(this._a, this._r, this._g, this._b);
	}
}

File: dxDrawLib/src/Server/Helpers/Color.cs
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;

namespace dxDrawLib.Server.Helpers
{
    public class Color
    {

        [JsonProperty("a")]
        private int _a;
        [JsonProperty("r")]
        private int _r;
        [JsonProperty("g")]
        private int _g;
        [JsonProperty("b")]
        private int _b;


        public Color(string hex)
        {
            if (hex.StartsWith("#")) hex = hex.Substring(1);
            int length = hex.Length;
            
            long value = Convert.ToInt32(hex, 16);

            if (length == 8) this._a = (int) (value >> 8 * 3) & 255;
            else this._a = 255;
            
            this._r = (int) (value >> 8 * 2) & 255;
            this._g = (int) (value >> 8 * 1) & 255;
            this._b = (int) value & 255;
        }
        
        public Color(int a, int r, int g, int b)
        {
            this.a = a;
            this.r = r;
            this.g = g;
            this.b = b;
        }
        
        public Color(int r, int g, int b)
        {
            this.a = 255;
            this.r = r;
            this.g = g;
            this.b = b;
        }
        
        [JsonIgnore]
        public int a
        {
            set { this._a = Helpers.Clamp(value, 0, 255); }
            get { return this._a; }
        }

        [JsonIgnore]
        public int r
        {
            set { this._r = Helpers.Clamp(value, 0, 255); }
            get { return this._r; }
        }

        [JsonIgnore]
        public int g
        {
            set { this._g = Helpers.Clamp(value, 0, 255); }
            get { return this._g; }
        }

        [JsonIgnore]
        public int b
        {
            set { this._b = Helpers.Clamp(value, 0, 255); }
            get { return this._b; }
        }

        public void Lighten(float factor)
        {
            this.r = (int) (this._r + (255 - this._r) * factor);
            this.g = (int) (this._g + (255 - this._g) * factor);
            this.b = (int) (this._b + (255 - this._b) * factor);
        }

        public void Darken(float factor)
        {
            this.r = (int) (this._r * (1 - factor));
            this.g = (int) (this._g * (1 - factor));
            this.b = (int) (this._b * (1 - factor));
        }

        public Color Clone() => new Color(this._a, this._r, this._g, this._b);

    }
}